specs/networking/p2p-interface.md

```diff
@@ -0,0 +1,743 @@
+
+1. If `fork_version` doesn’t match the local fork version, since the client’s chain is on another fork. `fork_version` can also be used to segregate testnets.
+2. If the (`finalized_root`, `finalized_epoch`) shared by the peer is not in the client's chain at the expected epoch. For example, if Peer 1 sends (root, epoch) of (A, 5) and Peer 2 sends (B, 3) but Peer 1 has root C at epoch 3, then Peer 1 would disconnect because it knows that their chains are irreparably disjoint.
+
+Once the handshake completes, the client with the lower `finalized_epoch` or `head_slot` (if the clients have equal `finalized_epoch`s) SHOULD request beacon blocks from its counterparty via the `BeaconBlocks` request.
+
+#### Goodbye
+
+**Protocol ID:** ``/eth2/beacon_chain/req/goodbye/1/``
+
+**Content:**
+```
+(
+  reason: uint64
+)
+```
+Client MAY send goodbye messages upon disconnection. The reason field MAY be one of the following values:
+
+- 1: Client shut down.
+- 2: Irrelevant network.
+- 3: Fault/error.
+
+Clients MAY use reason codes above `128` to indicate alternative, erroneous request-specific responses.
+
+The range `[4, 127]` is RESERVED for future usage.
+
+#### BeaconBlocks
+
+**Protocol ID:** `/eth2/beacon_chain/req/beacon_blocks/1/`
+
+Request Content
+```
+(
+  head_block_root: HashTreeRoot
+  start_slot: uint64
+  count: uint64
+  step: uint64
+)
+```
+
+Response Content:
+```
+(
+  blocks: []BeaconBlock
+)
+```
+
+Requests count beacon blocks from the peer starting from `start_slot` on the chain defined by `head_block_root`. The response MUST contain no more than count blocks. step defines the slot increment between blocks. For example, requesting blocks starting at `start_slot` 2 with a step value of 2 would return the blocks at [2, 4, 6, …]. In cases where a slot is empty for a given slot number, no block is returned. For example, if slot 4 were empty in the previous example, the returned array would contain [2, 6, …]. A step value of 1 returns all blocks on the range `[start_slot, start_slot + count)`.
+
+`BeaconBlocks` is primarily used to sync historical blocks.
+
+Clients MUST support requesting blocks since the start of the weak subjectivity period and up to the given `head_block_root`.
+
+Clients MUST support `head_block_root` values since the latest finalized epoch.
+
+#### RecentBeaconBlocks
+
+**Protocol ID:** `/eth2/beacon_chain/req/recent_beacon_blocks/1/`
+
+Request Content:
+
+```
+(
+  block_roots: []HashTreeRoot
+)
+```
+
+Response Content:
+
+```
+(
+  blocks: []BeaconBlock
+)
+```
+
+Requests blocks by their block roots. The response is a list of `BeaconBlock` with the same length as the request. Blocks are returned in order of the request and any missing/unknown blocks are left empty (SSZ null `BeaconBlock`).
+
+`RecentBeaconBlocks` is primarily used to recover recent blocks, for example when receiving a block or attestation whose parent is unknown.
+
+Clients MUST support requesting blocks since the latest finalized epoch.
+
```

Comment:
Do we need/want to request recent beacon state(s)? to eg: fetch a latest finalized state for syncing initially

Reply:
State sync is something we certainly should add, but we need to examine a set of tradeoffs along with the current state sync upgrade proposals currently being considered in eth1. 

Fortunately for interop and even for mainnet for a while, we will be fine with a pure block sync. That said, I think we should have something in place or at least a proposal in consideration before mainnet launch

Reply:
agreed, post-interop at least.